charms/kserve-controller/src/charm.py

```diff
@@ -476,14 +512,35 @@ def _on_event(self, event):
             # configuration is changed, otherwise the service will remain
             # unaware of such changes.
             self._restart_controller_service()
+
+            # If the Pod is not ready (condition with type Ready, all containers must be Ready)
+            # then K8s will drop request to svc with message "connect: connection refused".
+            # The charm container will become ready only once the start event has completed, and
+            # the workload container's pebble readiness probes are healthy.
+            # Until then the Pod is not ready, non-ready containers, thus traffic will be dropped.
+            # https://github.com/canonical/kserve-operators/issues/301
+            try:
+                self.cluster_runtimes_resource_handler.apply()
+                self.model.unit.status = ActiveStatus()
+                log.info("KServe Controller Pod was ready. Applied all ClusterServingRuntimes.")
+            except ApiError as e:
+                if "connection refused" in e.status.message:
+                    msg = "Charm Pod is not ready yet. Will apply ClusterServingRuntimes later."
+                    log.warning(msg)
+                    self.model.unit.status = MaintenanceStatus(msg)
+                else:
+                    log.warning("Unexpected ApiError happened: %s", e)
+                    raise ErrorWithStatus(
+                        f"Unexpected ApiError happened: {e.status.message}",
+                        BlockedStatus,
+                    )
```

Comment:
question: Is there a reason this block is placed after restarting the controller? On contrary to the other handlers applying manifests

Reply:
yep I prefer if we keep a pattern for when we execute the apply of k8s resources handlers, IMO it simplifies debugging when needed

Reply:
Done!

https://github.com/canonical/kserve-operators/pull/314/commits/b93f9e37cd9a5d9559a2b9111dbd539fbd3a9ac6